File: acwing/算法提高课/第一章动态规划/1010.拦截导弹/1010.拦截导弹.cpp
#include <iostream>
#include <cstring>
#include <algorithm>

using namespace std;

const int N = 1010;

int f[N], a[N], g[N];

int main()
{

    int n = 0;
    while (cin >> a[++n])
        ;
    n--;
    int res = 0;
    for (int i = 1; i <= n; i++)
    {
        f[i] = 1;
        for (int j = 1; j < i; j++)
        {
            if (a[i] <= a[j])
                f[i] = max(f[i], f[j] + 1);
        }
        res = max(res, f[i]);
    }
    printf("%d\n", res);

    int cnt = 0;
    for (int i = 1; i <= n; i++)
    {
        int k = 0;
        while (k < cnt && g[k] < a[i])
        {
            k++;
        }
        g[k] = a[i];
        if (k >= cnt)
            cnt++;
    }
    printf("%d\n", cnt);

    return 0;
}

File: acwing/算法提高课/第一章动态规划/1010.拦截导弹/1010.拦截导弹.py
N = 1010

f = [0] * N
g = [0] * N

a = []


a = [0] + list(map(int, input().split()))

n = len(a) - 1

cnt = 0


for i in range(n, 0, -1):
    f[i] = 1
    for j in range(n, i, -1):
        if a[i] >= a[j]:
            f[i] = max(f[i], f[j] + 1)
    cnt = max(cnt, f[i])

print(cnt)

res = 0

for i in range(1, n + 1):
    k = 0
    while k < res and g[k] < a[i]:
        k += 1
    g[k] = a[i]
    if k >= res:
        res += 1

print(res)
